services/ingestion/subscriber.go

```diff
@@ -2,55 +2,232 @@ package ingestion
-	transactionExecutedEvent := common.NewAddressLocation(nil, evmAddress, string(types.EventTypeTransactionExecuted)).ID()
+// Subscribe will retrieve all the events from the provided height. If the height is from previous
+// sporks, it will first backfill all the events in all the previous sporks, and then continue
+// to listen all new events in the current spork.
+//
+// If error is encountered during backfill the subscription will end and the response chanel will be closed.
+func (r *RPCSubscriber) Subscribe(ctx context.Context, height uint64) <-chan models.BlockEvents {
+	events := make(chan models.BlockEvents)
 
-	filter := flow.EventFilter{
-		EventTypes: []string{
-			blockExecutedEvent,
-			transactionExecutedEvent,
-		},
-	}
+	go func() {
+		defer func() {
+			close(events)
+		}()
+
+		// if the height is from the previous spork, backfill all the events from previous sporks first
+		if r.client.IsPastSpork(height) {
+			r.logger.Info().
+				Uint64("height", height).
+				Msg("height found in previous spork, starting to backfill")
+
+			for ev := range r.backfill(ctx, height) {
+				events <- ev
+
+				if ev.Err != nil {
+					return
+				}
+
+				// keep updating height, so after we are done back-filling
+				// it will be at the first height in the current spork
+				height = ev.Events.CadenceHeight()
+			}
+
+			// after back-filling is done, increment height by one,
+			// so we start with the height in the current spork
+			height = height + 1
+		}
+
+		r.logger.Info().
+			Uint64("next-height", height).
+			Msg("backfilling done, subscribe for live data")
+
+		// subscribe in the current spork
+		for ev := range r.subscribe(ctx, height) {
+			events <- ev
+		}
+
+		r.logger.Warn().Msg("ended subscription for events")
+	}()
+
+	return events
+}
+
+// subscribe to events by the provided height and handle any errors.
+func (r *RPCSubscriber) subscribe(ctx context.Context, height uint64, opts ...grpc.SubscribeOption) <-chan models.BlockEvents {
+	events := make(chan models.BlockEvents)
 
 	_, err := r.client.GetBlockByHeight(ctx, height)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to subscribe for events, the block height %d doesn't exist: %w", height, err)
+		err = fmt.Errorf("failed to subscribe for events, the block height %d doesn't exist: %w", height, err)
+		events <- models.NewBlockEventsError(err)
+		return events
 	}
 
-	// todo revisit if we should use custom heartbeat interval grpc.WithHeartbeatInterval(1)
-	evs, errs, err := r.client.SubscribeEventsByBlockHeight(ctx, height, filter)
+	evs, errs, err := r.client.SubscribeEventsByBlockHeight(ctx, height, r.blocksFilter(), opts...)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to subscribe to events by block height: %w", err)
+		events <- models.NewBlockEventsError(fmt.Errorf("failed to subscribe to events by block height: %w", err))
+		return events
 	}
 
-	return evs, errs, nil
+	go func() {
+		defer func() {
+			close(events)
+		}()
+
+		for {
```

Comment:
consider this to ensure you catch context cancellation. otherwise there's a race with the other channels
```suggestion
		for ctx.Err() == nil {
```

Reply:
nice yeah